src/services/refactors/convertToOptionalChainExpression.ts

```diff
@@ -0,0 +1,268 @@
+        if (isIdentifier(source) && isIdentifier(target)) {
+            return source.getText() === target.getText();
+        }
+        else if (isPropertyAccessExpression(source) && isPropertyAccessExpression(target)) {
+            return source.name.getText() === target.name.getText();
+        }
+        return false;
+    }
+
+    /**
+     * Find the least ancestor of the input node that is a valid type for extraction and contains the input span.
+     */
+    function getValidParentNodeContainingSpan(node: Node, span: TextSpan): ValidExpressionOrStatement | undefined {
+        while (node.parent) {
+            if (isValidExpressionOrStatement(node) && span.length !== 0 && node.end >= span.start + span.length) {
+                return node;
+            }
+            node = node.parent;
+        }
+        return undefined;
+    }
+
+    /**
+     * Finds an ancestor of the input node that is a valid type for extraction, skipping subexpressions.
+     */
+    function getValidParentNodeOfEmptySpan(node: Node): ValidExpressionOrStatement | undefined {
+        while (node.parent) {
+            if (isValidExpressionOrStatement(node) && !isValidExpressionOrStatement(node.parent)) {
+                return node;
+            }
+            node = node.parent;
+        }
+        return undefined;
+    }
+
+    /**
+     * Gets an expression of valid extraction type from a valid statement or expression.
+     */
+    function getExpression(node: ValidExpressionOrStatement): ValidExpression | undefined {
+        if (isValidExpression(node)) {
+            return node;
+        }
+        if (isVariableStatement(node)) {
+            const variable = getSingleVariableOfVariableStatement(node);
+            const initializer = variable?.initializer;
+            return initializer && isValidExpression(initializer) ? initializer : undefined;
+        }
+        return node.expression && isValidExpression(node.expression) ? node.expression : undefined;
+    }
+
+    /**
+     * Gets a property access expression which may be nested inside of a call expression or binary expression. The final
+     * expression in an && chain will occur as the right child of the parent binary expression, unless it is a call expression
+     * or is followed by a different binary operator.
+     * @param node the right child of a binary expression or a call expression.
+     */
+    function getLastPropertyAccessChain(node: Expression): PropertyAccessExpression | undefined {
+        // foo && |foo.bar === 1|; - here the right child of the && binary expression is another binary expression.
+        // the rightmost member of the && chain should be the leftmost child of that expression.
+        if (isBinaryExpression(node)) {
+            return getLastPropertyAccessChain(node.left);
+        }
+        // foo && |foo.bar()()| - if the right child is a call expression, simply search its expression.
+        else if (isCallExpression(node)) {

```

Comment:
as you told me earlier, probably this needs to be tweaked so that `a,ab,ab()` are `a,ab,ab()()` are treated the same as `a,ab,abc` or `a,ab,abcd` respectively, meaning that they need to return `ab()` and `ab()()` from here, respectively.

Then `getBinaryInfo` will have to handle both PropertyAccess and CallExpressions, which shouldn't be too hard. And it looks like getMatchingSubexpression already handles CallExpressions.

Reply:
We now return the call expression and skip past it along with "trailing" property accesses in the subchain prefix check.